### DISCUSSION TITLE
Control auto completion suggestions

### DISCUSSION THREAD

**Discussion by paulfrench** [eclipse-langium/langium#384] [General]
I'm only showing the relevant parts of the grammar file...

```
Entity:
	'entity' name=ID ('extends' superType = [Entity])? '{'
		(properties+=Property)*
    ('keys' ':' '[' keys+=[Property] ("," keys+=[Property])* ']' )?
	'}';

Query:
	'query' name=ID '{'
		'find' ':' entity=[Entity] 'for' properties+=[Property] (("," properties+=[Property])*)?
	'}';
```
What I would like to do is for the **query find** part, where after specifying the Entity, only show a list of properties to choose from by the specified **[Entity]** ? I know the [Property] reference I have put in the Query rule will not work since it is not in the context of the specified [Entity] (unlike for the Entity rule where the keys suggested are the properties of the Entity)

So, is it possible out of the box or is this more complex where you need to provide your own CompletionProvider? If so I've looked at the DefaultCompletionProvider and it looks fairly complex so any pointers would be really good.

Cheers

**Comment by debovema** [comment]
Interested in an example to subclass DefaultCompletionProvider

**Reply by paulfrench** [reply]
Your interested or am I interested? Not sure what you mean?

**Reply by debovema** [reply]
I am also interested.

[Documentation](https://langium.org/docs/configuration-services/#language-server-protocol) says that we can "modify aspects of the Language Server" but I don't succeed to configure my own CompletionProvider.

**Comment by msujew** [comment]
@paulfrench @debovema Sorry for coming back to this so late, seems like I forgot about this.

Anyway, regarding the actual question here: First off, there is a bug which prevents Langium from correctly presenting completion results for cross references which are missing their terminal (i.e. modify `[Property]` to `[Property:ID]`) See also [here](https://github.com/langium/langium/discussions/401#discussioncomment-2388207). Secondly, you shouldn't need to modify the completion provider for this. Instead, your scoping should "redirect" the cross reference to the `Property` element to the properties provided by the referenced `Entity`. Our completion works based on scoping, so that should automatically propagate to completion.

On another note: To override the completion provider you just need to override the `LangiumService.lsp.completion.CompletionProvider` module entry. Something like this:

```ts
class CustomCompletionProvider extends DefaultCompletionProvider {
   ... // Your custom overrides here
}

// In your module.ts
export const CustomModule: Module<CustomServices, PartialLangiumServices & CustomAddedServices> = {
    lsp: {
        completion: {
            CompletionProvider: (services) => new CustomCompletionProvider(services)
        }
    }
};
```
